[마켓불리 김원희] Add: 고객 후기(리뷰) Endpoint 구현 완료

supersaloon opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- 상품 상세페이지 하단에 출력되는 고객 후기 게시판 Endpoint 구현 완료
		 
## 수정 사항들 자세한 내용
상품 상세 페이이지 하단에 출려되는 고객 후기 구현 완료
  - 후기 조회 구현 완료
    - 사용자들이 고객 후기를 조회시 hit_count += 1 하여 조회수 증가시켜 저장
  - 후기 수정 구현 완료
    - 사용자가 자신이 작성한 후기에 대해서 patch 메소드로 수정 가능
    - 다른 사용작 후기에 "도움" 버튼을 클릭한 경우  
      - patch 메소드로 다른 키 값 없이, `help="True"`로 request 하면 help_count 만 증가시킴 
  - 후기 삭제 구현 완료
    - 사용자 자신이 작성한 후기에 대해서 삭제 가능
  - 후기 목록 조회 구현 완료  
    - 프론트에서 전달받은 product_id로 조회하여 -id(id 역순)으로 후기를 정렬하고  
    - API query sting으로 전달받은 offset, limit으로 목록을 가공하여 프론트에 전달

## 기타 질문 및 특이 사항
- 고객 후기 목록 출력 부분의 에러처리 관련하여 질문드립니다.
- 다른 부분 구현시엔 전달 받은 키 값으로 인해 발생되는 에러에 대해서 다음과 같은 방식으로 처리했었습니다.
```python
try:
    data = json.loads(request.body)
    something = data['something']
except KeyError as e:
    return JsonResponse({"MESSAGE": "KEY_ERROR => " + e.args[0]}, status=400)
```
- 현재 고객 후기 목록의 에러처리를 아래와 같이 구현하였는데, 별로 좋은 방법이 아닌걸로 알고 있습니다.
- 혹시 쿼리스트링으로 Key-Value를 전달 받는 경우에 좀더 세련된 접근 방법이 있을까요?
```python
class ReviewListView(View):
    # @login_required
    def get(self, request, product_id):
        try:
            offset = int(request.GET.get('offset'), 0)
            limit  = int(request.GET.get('limit'), 10)
            limit += offset

            reviews = Review.objects.order_by('-id').filter(product_id=product_id)

            review_list = [{
                'id'        : review.id,
                'title'     : review.title,
                'contents'  : review.contents,
                'help_count': review.help_count,
                'hit_count' : review.hit_count,
                'image_url' : review.image_url,

            }for review in reviews[offset:limit]]

            return JsonResponse({'MESSAGE': 'SUCCESS', "review_list": review_list}, status=200)

        except Exception as e:
            return JsonResponse({'message' : 'ERROR => ' + e.args[0]}, status = 400)
```


## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
Add: 사용자 후기 구현 완료
- 조회시 hit_count += 1
- 후기 수정 구현 완료
- 사용자가 후기에 대해 수정한 경우 patch 로 수정 가능
- 다른 사용작 후기에 "도움" 버튼을 클릭한 경우
  - patch 메소드로 다른 키 값 없이, `help="True"`로 request 하면 help_count 만 증가시킴
- 후기 삭제 구현 완료
- 후기 목록 조회 구현 완료
  - product_id로 조회하여 -id(id 역순)으로 후기를 정렬하고
  - API query sting으로 offset, limit으로 가공한 목록을 프론트에 전달

Commit:
board/urls.py 수정

soheon-lee commented on board/views.py:
```diff
@@ -1,3 +1,112 @@
+            limit  = int(request.GET.get('limit'), 10)
+            limit += offset
+
+            reviews = Review.objects.order_by('-id').filter(product_id=product_id)
```
혹시 리뷰는 최신 생성순으로 나열하는 것이 더 좋지 않을까요?

supersaloon replied:
최신 생성순(-create_at)기준으로 출력되도록 수정하였습니다.

soheon-lee commented on board/views.py:
```diff
@@ -1,3 +1,112 @@
+            return JsonResponse({"MESSAGE": "KEY_ERROR => " + e.args[0]}, status=400)
+        except Review.DoesNotExist:
+            return JsonResponse({"MESSAGE": "REVIEW_DOES_NOT_EXIST"}, status=400)
+
+    # @login_required
+    def delete(self, request, review_id):
+        try:
+            user_id = "3"
+
```
원희님 user 부분이 merge된걸로 알고 있는데, user_id 가져오는 부분 수정하셔야 할 것 같습니다 :)

supersaloon replied:
기존 하드코딩된 부분 제거하고, 데코레이터로 인증 받도록 수정하였습니다.

soheon-lee commented on board/views.py:
```diff
@@ -1,3 +1,112 @@
+            review = Review.objects.get(id=review_id)
+            review.hit_count += 1
+            review.save()
+            review_post = [{
+                'id'        : review.id,
+                'title'     : review.title,
+                'contents'  : review.contents,
+                'help_count': review.help_count,
+                'hit_count' : int(review.hit_count),
+                'image_url' : review.image_url,
+            }]
```
리뷰가 하나 뿐이라 list에 담지 않으셔도 될 것 같습니다

supersaloon replied:
리스트에 담지 않도록 수정하였습니다.

Commit:
Modify: 고객 리뷰 기능에 로그인 데코레이터 적용
- 리뷰 작성시 데코레이터 적용하여 로그인한 사용자만 가능하도록 구현
- 리뷰 삭제시 데코레이터 적용하여 로그인한 사용자만 가능하도록 구현
- 리뷰 수정시 데코레이터 적용하여 로그인한 사용자만 가능하도록 구현

Commit:
Modify: 리뷰 상세 출력시 리스트에 담지 않고 바로 딕셔너리에 담도록 수정, 리뷰 목록 출력시 최신순(-created_at) 기준으로 출력되도록 수정

supersaloon committed:
Merge branch 'master' into feature/board-reviews

Commit:
Modify: 리뷰 목록에 생성일 출력되도록 수정

Commit:
Merge branch 'feature/board-reviews' of https://github.com/wecode-bootcamp-korea/15-1st-MarketBully-backend into feature/board-reviews

Commit:
Modify: 리뷰 상세보기에서 create_at 출력

jamessoun93 commented:
LGTM
